### ISSUE TITLE
Multi Worker Mirrored Strategy -- math

### ISSUE THREAD

**Issue by funmathgt** [tensorflow/tensorflow#47676]
In the Multi Worker Mirrored Strategy (seen here https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras) it is not clear what is going on with the gradients and backpropogation under the hood (i.e., the math). For example, are gradients being averaged among the workers in the cluster? I cannot seem to figure this out as when I run on one machine without using a distributed strategy and on GPU my training/validation loss and training/validation metric differ significantly when I run on a cluster of machines using the distributed strategy Multi Worker Mirrored Strategy.  I was hoping this would not happen, i.e., you would add machines to the cluster to ideally speed up training, with loss and performance metric (MSE, RMSE) being roughly equal.

Also, if there is some reading/documentation on the math/how this is being implemented, I would appreciate. it. Could not find any so far.

**Comment by amahendrakar** [comment]
@inespancorbo,
Could you please go through [this guide](https://www.tensorflow.org/guide/distributed_training#multiworkermirroredstrategy) which gives a brief overview of MultiWorkerMirroredStrategy and let us know if it helps. Thanks!

**Comment by inespancorbo** [comment]
I have already looked at that unfortunately. I am asking about gradient updates/losses etc. How are these handled when many machines are involved?

**Comment by nikitamaia** [comment]
Hi @inespancorbo, I think the mental model here is easier to understand if we start with the `MirroredStrategy` case.

`MirroredStrategy` is a synchronous data parallelism strategy. This means that when you call `model.fit`, `MirroredStrategy` will make a copy (known as a replica) of your model on each GPU. The CPU (host) is responsible for preparing the `tf.data.Dataset` batches and sending the data to the GPUs (devices).

The subsequent gradient updates will happen in a synchronous manner. This means that each replica computes the forward and backward passes through the model on a different slice of the input data. 

The computed gradients from each of these slices are then aggregated across all of the devices and reduced (usually an average) in a process known as all-reduce. There are many all-reduce algorithms and implementations available, depending on the type of communication available between devices. By default, `MirroredStrategy` uses [NVIDIA NCCL](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/index.html) as the all-reduce implementation.

After the reduction, the optimizer then performs the parameter updates with these reduced gradients thereby keeping the devices in sync. Because each worker cannot proceed to the next training step until all the other workers have finished the current step, this gradient calculation becomes the main overhead in distributed training for synchronous strategies.

So that's what happens in the case of a single machine with multiple GPUs.

Abstracting out to multiple machines is not much different. Like its single-worker counterpart, `MirroredStrategy`, `MultiWorkerMirroredStrategy` is a synchronous data parallelism strategy.

The main difference when moving from synchronous data parallelism on one machine to many machines is that the gradients at the end of each step now need to be synchronized across all GPUs in a machine and across all machines in the cluster. This additional step of synchronizing across the machines increases the overhead of distribution. 

In TensorFlow, the multi-worker all-reduce communication is achieved via CollectiveOps. You don’t need to know much detail to execute a successful and performant training job, but at a high level,  a collective op is a single op in the TensorFlow graph that can automatically choose an all-reduce algorithm according to factors such as hardware, network topology, and tensor sizes.

The last thing to note is the distribution of the data. In the single worker case, at each step your dataset is divided up across the replicas on your machine. So when you do distributed training with the `tf.distribute.Strategy` API and `tf.data`, the batch size now refers to the global batch size. In other words, if you pass a batch size of 10, and you have two GPUs, then each machine will process 5 examples per step. In this case, 10 is known as the _global batch size_, and 5 as the per replica batch size. To make the most out of your GPUs, you will want to scale the batch size by the number of replicas, which is two in this case because there is one replica on each GPU.

```
GLOBAL_BATCH_SIZE = PER_REPLICA_BATCH_SIZE * strategy.num_replicas_in_sync
```

This data splitting process becomes slightly more complicated in the multi-worker case. The data now also needs to be sharded, meaning that each worker is assigned a subset of the entire dataset. Therefore, at each step a global batch size of non overlapping dataset elements will be processed by each worker. This sharding happens automatically with `tf.data.experimental.AutoShardPolicy.` You can read more about the different sharding policies and see examples in the [Distributed Input guide](https://www.tensorflow.org/tutorials/distribute/input).

Hopefully this explanation at least helps with the mental model of what's going on and gives a starting point if you'd like to dig into the math details.

I'm closing this issue now because Github is best for bugs/performance issues. For further support on this topic, I recommend posting in Stack Overflow. And if you think you've discovered a bug, please open a new issue following the bug template. Thanks!
